### PR TITLE
Introduce `Conditional` validator

### LINKED ISSUE #370: New conditional validator
The idea is to add a new validator that executes a validator when a rule is met.

## Example

```php
$validator = new Laminas\Validator\ConditionalValidator(
    [
        'validator' => [
            'name' => Laminas\Validator\NotEmpty::class,
        ],
        'rule'      => static function (array $context) {
            return (bool) ($context['confirm'] ?? false);
        },
    ]
);
```

```php
$data = [
    'registration' => 'foo',
    'confirm'      => '1',
];

var_dump($validator->isValid($data['registration'], $data)); // true

$data = [
    'registration' => '',
    'confirm'      => '1',
];

var_dump($validator->isValid($data['registration'], $data)); // false

$data = [
    'registration' => '',
    'confirm'      => '',
];

var_dump($validator->isValid($data['registration'], $data)); // true
```

## Implementation

The validator needs the plugin manager and a solution to allow a chain of validators, for example a third option – `validators` next to `validator`:

```php
$validator = new Laminas\Validator\ConditionalValidator(
    [
        'validators' => [
            [
                'name'    => Laminas\Validator\StringLength::class,
                'options' => [
                    'min' => 3,
                    'max' => 255,
                ],
            ],
            [
                'name' => Laminas\I18n\Validator\Alnum::class,
            ],
        ],
        'rule'      => static function (array $context) {
            return (bool) ($context['confirm'] ?? false);
        },
    ]
);
```

## Reference

https://book.cakephp.org/5/en/core-libraries/validation.html#conditional-validation

### PR DESCRIPTION
**Pull request by gsteel**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| New Feature   | yes

### Description

Closes #370

### CODE REVIEW

File: docs/book/v3/validators/conditional.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+- `rule`: A callable that will determine whether validation should execute on the value or not
+- `validators`: A list of validator specifications used to create a [validator chain](../validator-chains.md)
+
+## About the `$context` parameter of validators
+
+Typically, `laminas-validator` is used via `laminas-inputfilter` which is often, in turn, used via `laminas-form`.
+Some validators accept a second parameter to the `isValid()` method that contains the entire payload in an unfiltered and un-validated state.
+This parameter `$context` is normally the entire `$_POST` payload.
+
+`laminas-inputfilter` always passes this parameter to the `isValid` method, but, because it is not part of the `ValidatorInterface` contract, it's documentation has often been overlooked.
```
Comment (froschdesign) [inline docs/book/v3/validators/conditional.md:18]: This section is important and the parameter is also used in the following classes:

- Callback
- Explode
- Identical
- ValidatorChain

At least for the Callback validator, the section must be adopted, for all others it is internal processing.

See also: #96

File: docs/book/v3/validators/conditional.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+Normally, using validators via `laminas-inputfilter` or `laminas-form`, this will not be a concern, but in standalone usage as documented here, we must account for this requirement.
+
+In the following example, we only wish to validate the email address if the `subscribe` field evaluates to a truthy value:
+
+```php
+use Laminas\Validator\Conditional;
+use Laminas\Validator\EmailAddress;
+use Laminas\Validator\ValidatorChainFactory;
+
+$chainFactory = $container->get(ValidatorChainFactory::class);
```
Comment (froschdesign) [inline docs/book/v3/validators/conditional.md]: We already had this problem in the documentation of laminas-cache, so I propose to show the creation of the container:

```suggestion
$container = new Laminas\Validator\ValidatorPluginManager(
    new Laminas\ServiceManager\ServiceManager()
);
$chainFactory = $container->get(ValidatorChainFactory::class);
```

**Review by froschdesign — CHANGES_REQUESTED** [review]

File: docs/book/v3/validators/conditional.md (hunk 1 above)
Comment (gsteel) [reply docs/book/v3/validators/conditional.md:18]: This has been documented in the Callback validator and CreditCard validator which composes a Callback validator in #379 … I'll do a separate patch for the other docs.

File: docs/book/v3/validators/conditional.md (hunk 2 above)
Comment (gsteel) [reply docs/book/v3/validators/conditional.md]: The chain factory is retrieved from the main DI container, not the plugin manger, so I've instantiated it directly in the docs instead.

Comment (froschdesign) [reply docs/book/v3/validators/conditional.md]: Of course you're absolutely right!

**Review by froschdesign — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**